File: python/check_domain.py
#!/usr/bin/env python3
"""
TECHli Email Security API - Python Example
Check a single domain's email security configuration.

Usage:
    python check_domain.py example.com
    python check_domain.py example.com --json
"""

import sys
import json
import requests
from typing import Dict, Any


API_URL = "https://techli.nz/api/test-domain"


def check_email_security(domain: str) -> Dict[str, Any]:
    """
    Check email security for a domain using TECHli API.
    
    Args:
        domain: Domain name to check (e.g., 'example.com')
        
    Returns:
        Dictionary containing test results
        
    Raises:
        requests.RequestException: If API request fails
    """
    response = requests.post(
        API_URL,
        json={"domain": domain},
        headers={"Content-Type": "application/json"},
        timeout=10
    )
    response.raise_for_status()
    return response.json()


def format_status(status: str) -> str:
    """Format status with color emoji."""
    status_map = {
        "pass": "✅ PASS",
        "fail": "❌ FAIL",
        "warning": "⚠️  WARNING"
    }
    return status_map.get(status, status.upper())


def print_report(result: Dict[str, Any]) -> None:
    """Print a formatted security report."""
    print(f"\n{'='*60}")
    print(f"  EMAIL SECURITY REPORT: {result['domain']}")
    print(f"{'='*60}\n")
    
    # Overall score
    score = result['overallScore']
    compliance = result['complianceLevel'].upper()
    print(f"Overall Score: {score}/100")
    print(f"Compliance: {compliance}\n")
    
    # SPF
    print(f"SPF:      {format_status(result['spf']['status'])}")
    print(f"          {result['spf']['message']}")
    if result['spf'].get('record'):
        print(f"          Record: {result['spf']['record']}")
    print()
    
    # DKIM
    print(f"DKIM:     {format_status(result['dkim']['status'])}")
    print(f"          {result['dkim']['message']}")
    if result['dkim'].get('selector'):
        print(f"          Selector: {result['dkim']['selector']}")
    print()
    
    # DMARC
    print(f"DMARC:    {format_status(result['dmarc']['status'])}")
    print(f"          {result['dmarc']['message']}")
    if result['dmarc'].get('policy'):
        print(f"          Policy: {result['dmarc']['policy']}")
    print()
    
    # MX
    print(f"MX:       {format_status(result['mx']['status'])}")
    print(f"          {result['mx']['message']}")
    for mx in result['mx'].get('records', []):
        print(f"          [{mx['priority']}] {mx['exchange']}")
    print()
    
    # BIMI
    if result.get('bimi'):
        print(f"BIMI:     {format_status(result['bimi']['status'])}")
        print(f"          {result['bimi']['message']}")
        if result['bimi'].get('logoUrl'):
            print(f"          Logo: {result['bimi']['logoUrl']}")
        print()
    
    # MTA-STS
    if result.get('mtasts'):
        print(f"MTA-STS:  {format_status(result['mtasts']['status'])}")
        print(f"          {result['mtasts']['message']}")
        if result['mtasts'].get('mode'):
            print(f"          Mode: {result['mtasts']['mode']}")
        print()
    
    print(f"{'='*60}")
    print(f"Tested at: {result['timestamp']}")
    print(f"{'='*60}\n")


def main():
    """Main function."""
    if len(sys.argv) < 2:
        print("Usage: python check_domain.py <domain> [--json]")
        print("Example: python check_domain.py example.com")
        sys.exit(1)
    
    domain = sys.argv[1]
    output_json = "--json" in sys.argv
    
    try:
        print(f"Checking email security for: {domain}...")
        result = check_email_security(domain)
        
        if output_json:
            print(json.dumps(result, indent=2))
        else:
            print_report(result)
            
    except requests.RequestException as e:
        print(f"❌ Error: Failed to check domain: {e}", file=sys.stderr)
        sys.exit(1)
    except KeyError as e:
        print(f"❌ Error: Unexpected API response format: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: nodejs/check-domain.js
#!/usr/bin/env node
/**
 * TECHli Email Security API - Node.js Example
 * Check a single domain's email security configuration.
 * 
 * Usage:
 *   node check-domain.js example.com
 *   node check-domain.js example.com --json
 */

const https = require('https');

const API_URL = 'https://techli.nz/api/test-domain';

/**
 * Check email security for a domain
 */
async function checkEmailSecurity(domain) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify({ domain });
    
    const options = {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': data.length
      }
    };
    
    const req = https.request(API_URL, options, (res) => {
      let body = '';
      
      res.on('data', (chunk) => {
        body += chunk;
      });
      
      res.on('end', () => {
        try {
          const result = JSON.parse(body);
          resolve(result);
        } catch (e) {
          reject(new Error('Invalid JSON response'));
        }
      });
    });
    
    req.on('error', reject);
    req.write(data);
    req.end();
  });
}

/**
 * Format status with emoji
 */
function formatStatus(status) {
  const statusMap = {
    'pass': '✅ PASS',
    'fail': '❌ FAIL',
    'warning': '⚠️  WARNING'
  };
  return statusMap[status] || status.toUpperCase();
}

/**
 * Print formatted report
 */
function printReport(result) {
  console.log('\n' + '='.repeat(60));
  console.log(`  EMAIL SECURITY REPORT: ${result.domain}`);
  console.log('='.repeat(60) + '\n');
  
  console.log(`Overall Score: ${result.overallScore}/100`);
  console.log(`Compliance: ${result.complianceLevel.toUpperCase()}\n`);
  
  // SPF
  console.log(`SPF:      ${formatStatus(result.spf.status)}`);
  console.log(`          ${result.spf.message}`);
  if (result.spf.record) {
    console.log(`          Record: ${result.spf.record}`);
  }
  console.log();
  
  // DKIM
  console.log(`DKIM:     ${formatStatus(result.dkim.status)}`);
  console.log(`          ${result.dkim.message}`);
  if (result.dkim.selector) {
    console.log(`          Selector: ${result.dkim.selector}`);
  }
  console.log();
  
  // DMARC
  console.log(`DMARC:    ${formatStatus(result.dmarc.status)}`);
  console.log(`          ${result.dmarc.message}`);
  if (result.dmarc.policy) {
    console.log(`          Policy: ${result.dmarc.policy}`);
  }
  console.log();
  
  // MX
  console.log(`MX:       ${formatStatus(result.mx.status)}`);
  console.log(`          ${result.mx.message}`);
  result.mx.records.forEach(mx => {
    console.log(`          [${mx.priority}] ${mx.exchange}`);
  });
  console.log();
  
  // BIMI
  if (result.bimi) {
    console.log(`BIMI:     ${formatStatus(result.bimi.status)}`);
    console.log(`          ${result.bimi.message}`);
    if (result.bimi.logoUrl) {
      console.log(`          Logo: ${result.bimi.logoUrl}`);
    }
    console.log();
  }
  
  // MTA-STS
  if (result.mtasts) {
    console.log(`MTA-STS:  ${formatStatus(result.mtasts.status)}`);
    console.log(`          ${result.mtasts.message}`);
    if (result.mtasts.mode) {
      console.log(`          Mode: ${result.mtasts.mode}`);
    }
    console.log();
  }
  
  console.log('='.repeat(60));
  console.log(`Tested at: ${result.timestamp}`);
  console.log('='.repeat(60) + '\n');
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.error('Usage: node check-domain.js <domain> [--json]');
    console.error('Example: node check-domain.js example.com');
    process.exit(1);
  }
  
  const domain = args[0];
  const outputJson = args.includes('--json');
  
  try {
    console.log(`Checking email security for: ${domain}...`);
    const result = await checkEmailSecurity(domain);
    
    if (outputJson) {
      console.log(JSON.stringify(result, null, 2));
    } else {
      printReport(result);
    }
  } catch (error) {
    console.error(`❌ Error: ${error.message}`);
    process.exit(1);
  }
}

main();
